### ISSUE TITLE
Deadlock with ProcessPoolExecutor + unpicklable work items + cancel_futures=True

### ISSUE THREAD

**Issue by dalleyg** [python/cpython#116685]
# Bug report

### Bug description:

Here is a simple reproducer of the problem. If the work items are not picklable (e.g. due to submitting lambdas), and there's an explicit call to the `shutdown` method with `cancel_futures=True`, then the `shutdown` call hangs.

```python
from concurrent.futures import ProcessPoolExecutor
import os
with ProcessPoolExecutor(1) as exe:
    futs = [exe.submit(lambda: 123) for _ in range(10)]
    try:
        print([fut.result() for fut in futs])
    finally:
        exe.shutdown(wait=True, cancel_futures=True)
```

I have tested this on CPython 3.9 on Ubuntu LTS and on CPython 3.11.8 using the official Python 3.11 Docker image.

```
docker run -it --rm --name py311-shell -v "$PWD":/usr/src/myapp -w /usr/src/myapp python:3.11 bash
root@<redacted>:/usr/src/myapp# python -c 'from concurrent.futures import ProcessPoolExecutor
import os
with ProcessPoolExecutor(1) as exe:
    futs = [exe.submit(lambda: 123) for _ in range(10)]
    try:
        print([fut.result() for fut in futs])
    finally:
        exe.shutdown(wait=True, cancel_futures=True)'
```

Traceback after hitting Ctrl+C:

```
CProcess ForkProcess-1:
concurrent.futures.process._RemoteTraceback:
"""
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/multiprocessing/queues.py", line 244, in _feed
    obj = _ForkingPickler.dumps(obj)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function <listcomp>.<lambda> at 0x7fef7060efc0>: attribute lookup <listcomp>.<lambda> on __main__ failed
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 6, in <module>
  File "<string>", line 6, in <listcomp>
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 456, in result
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
Traceback (most recent call last):
    self.run()
  File "/usr/local/lib/python3.11/concurrent/futures/process.py", line 344, in run
    result_item, is_broken, cause = self.wait_result_broken_or_wakeup()
    return self.__get_result()
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/process.py", line 417, in wait_result_broken_or_wakeup
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    ready = mp.connection.wait(readers + worker_sentinels)
  File "/usr/local/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.11/concurrent/futures/process.py", line 249, in _process_worker
    call_item = call_queue.get(block=True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/multiprocessing/queues.py", line 103, in get
    res = self._recv_bytes()
          ^^^^^^^^^^^^^^^^^^
    raise self._exception
  File "/usr/local/lib/python3.11/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/multiprocessing/connection.py", line 430, in _recv_bytes
    buf = self._recv(4)
          ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/multiprocessing/connection.py", line 395, in _recv
    chunk = read(handle, remaining)
            ^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/multiprocessing/connection.py", line 941, in wait
  File "/usr/local/lib/python3.11/multiprocessing/queues.py", line 244, in _feed
    obj = _ForkingPickler.dumps(obj)
    selector.register(obj, selectors.EVENT_READ)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/multiprocessing/reduction.py", line 51, in dumps
  File "/usr/local/lib/python3.11/selectors.py", line 352, in register
    cls(buf, protocol).dump(obj)
    key = super().register(fileobj, events, data)
_pickle.PicklingError: Can't pickle <function <listcomp>.<lambda> at 0x7fef7060efc0>: attribute lookup <listcomp>.<lambda> on __main__ failed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 8, in <module>
  File "/usr/local/lib/python3.11/concurrent/futures/process.py", line 851, in shutdown
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/selectors.py", line 238, in register
    self._executor_manager_thread.join()
    key = SelectorKey(fileobj, self._fileobj_lookup(fileobj), events, data)
  File "/usr/local/lib/python3.11/threading.py", line 1119, in join
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/selectors.py", line 225, in _fileobj_lookup
    self._wait_for_tstate_lock()
    return _fileobj_to_fd(fileobj)
  File "/usr/local/lib/python3.11/threading.py", line 1139, in _wait_for_tstate_lock
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/selectors.py", line 37, in _fileobj_to_fd
    if lock.acquire(block, timeout):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    fd = int(fileobj.fileno())
KeyboardInterrupt
             ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/multiprocessing/connection.py", line 171, in fileno
    self._check_closed()
  File "/usr/local/lib/python3.11/multiprocessing/connection.py", line 137, in _check_closed
```

### CPython versions tested on:

3.11

### Operating systems tested on:

Linux

**Comment by gaogaotiantian** [comment]
Seems like it's already fixed in 3.12. Can't repro it in 3.12 or main (able to repro in 3.11). If this is 3.11 only, I don't think we will fix it (as there was probably a reason why the fix was not backported when it was fixed).

**Comment by dalleyg** [comment]
I did some more testing. 

* Python 3.9.18, 3.10.13, 3.11.8: all hang (bad)
* Python 3.12.2: exits immediately (good)

Here's a reproducer for those results on Linux with Docker. It just requires pressing Ctrl+c on 3.9, 3.10, and 3.11 when they hang.
```
for V in 9 10 11 12; do
    echo "==== PYTHON $V ============================================================";
    docker run -it --rm -v "$PWD":/usr/src/myapp -w /usr/src/myapp python:3.$V \
        bash -c "echo 'from concurrent.futures import ProcessPoolExecutor
import os, sys; print(sys.version)
with ProcessPoolExecutor(1) as exe:
    futs = [exe.submit(lambda: 123) for _ in range(10)]
    try:
        print([fut.result() for fut in futs])
    finally:
        exe.shutdown(wait=True, cancel_futures=True)' > foo.py ; python foo.py"; 
    echo;    
done
```

Your suggestion sounds reasonable for now: don't worry about backporting the fix for < 3.12.
